### ISSUE TITLE
PageSections misbehave when locating elements with xpaths 

### ISSUE THREAD

**Issue by tk8817** [cheezy/page-object#354]
Hi @daneandersen,

I apologize in advance as I have not fully wrapped my head around this issue and thus may fumble in explaining it. I am working to recreate this failure outside of my framework so I can post it up for for you to replicate with.  I recently overhauled my framework from module includes (a la Cucumber & Cheese) to PageSections and they have been working great. I use them primarily to map collections of results as I spend most my time testing an e-commerce platform.  

__The issue:__ If you have a collection of PageSections and an element is located using an xpath, that element will potentially map to a section of HTML outside of the PageSections root. 

__The Consequence:__ In my example below, I map a list of stores using PageSections, some of which should have a 'SHOP THIS STORE' element, while others do not. However, if you look up the aformentioned element with an xpath, each page section will find the first element in the DOM which matches, and use that as the reference. So all of my PageSections are incorrect and end up referring to the first matched element in the DOM. 

__The Workaround:__ If you use any lookup other than xpath, then the PageSections work as expected and everything is fantastic. Unfortunately there are some key cases where an xpath is quite helpful. 

__The Example:__ Below I have a cucumber test case which should fail. It checks to see if all of the returned stores are 'shoppable'. In the code, this checks whether the 'select' element is visible in a given page section. When you lookup 'select' with a class, the test works and shows the appropriate stores (indexes) are missing this element. When you lookup 'select' with an xpath, the test passes!!! I did an experiment where I printed out the element.text and I found it was clearly using the same reference for all page sections.

__Webpage Being Tested:__ https://shop.loblaws.ca/store-locator

__Page Object Where Sections Are Created:__
``` ruby
# encoding: utf-8
class StoreLocator

  include PageObject
  include ::BasePage

  page_url ROOT_URL + FigNewton.partial_store_locator

  page_sections( :stores, StoreProperties,        class:  'store-locator-selector-list-item')

  span(         :store_locator_search_btn,        xpath:   './/div[@id="wrapper"]//span[@class="icon-search"]') #xpath as there are multiple items with this class
  text_field(   :store_locator_postal_code,          id:   'enter-new-search-term')
  span(         :store_locator_store_count,       class:   'num-of-stores')


  def initialize_page
    wait_for_ajax
  end

  def locate_stores(pcode)
    populate_page_with ({store_locator_postal_code: pcode})
    store_locator_search_btn_element.when_present.click
    wait_for_ajax

    Collector.log('store_lookup_data',
                  ["Postal Code: #{pcode}",
                   "Results: #{stores.count}"])
  end

end
```

__The Code Which Fails:__
- lookup element with an xpath
```ruby
class StoreProperties

  include PageObject

  h5(         :select,          xpath: '//h5[@class="store-actions shop-or-browse-store trigger-select-store"]')

end

  def shoppable?
    select_element.visible?
  end

###From Steps File###
Then(/^all stores should be shoppable$/) do
  stores = on(@current_page.class).stores
  expect(stores).to all(be_shoppable)
end
```

__Test Results (this should fail, but it passes)__
``` ruby
Feature: Test Scratchpad

  Scenario: Feature Setup
    Given I am a new visitor

  Scenario: Locate a set of stores
    When I open the store locator
    And I locate stores using the default postal code
    Then all stores should be shoppable

2 scenarios (2 passed)
5 steps (5 passed)
0m17.945s
```

__The Code Which Works:__
- only one change, lookup using class not xpath.

```ruby
class StoreProperties

  include PageObject

  h5(          :select,          class: 'trigger-select-store')

end

  def shoppable?
    puts select_element.text #just for debug purposes
    select_element.visible?
  end

###From Steps File###
Then(/^all stores should be shoppable$/) do
  stores = on(@current_page.class).stores
  expect(stores).to all(be_shoppable)
end
```

__Test Results:__
This works as expected, the test finds that some of the stores are not shoppable (there is no select store button present). 
 ```ruby
  Scenario: Feature Setup
    Given I am a new visitor

  Scenario: Locate a set of stores
    When I open the store locator
    And I locate stores using the default postal code
    Then all stores should be shoppable

      expected #<PageObject::SectionCollection:0x007fc3a381cd10 @sections=[#<StoreProperties:0x007fc3a30393f0 @root_element=#<Watir::HTMLElement:0x..fb80df99fbb14eb76 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a3032fc8 @browser=#<Watir::HTMLElement:0x..fb80df99fbb14eb76 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a3032f78 @root_element=#<Watir::HTMLElement:0x12ce9696a8cd3664 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a3031cb8 @browser=#<Watir::HTMLElement:0x12ce9696a8cd3664 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a3031c90 @root_element=#<Watir::HTMLElement:0x6a325a24f6ce0974 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a3031060 @browser=#<Watir::HTMLElement:0x6a325a24f6ce0974 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a3031038 @root_element=#<Watir::HTMLElement:0x42762409dfab5d52 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a30304f8 @browser=#<Watir::HTMLElement:0x42762409dfab5d52 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a30304d0 @root_element=#<Watir::HTMLElement:0x..f9ef1392241821c78 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a301b0a8 @browser=#<Watir::HTMLElement:0x..f9ef1392241821c78 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a301b080 @root_element=#<Watir::HTMLElement:0xdf0f6eb486a7010 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a3019140 @browser=#<Watir::HTMLElement:0xdf0f6eb486a7010 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a3019118 @root_element=#<Watir::HTMLElement:0x3298a7b0570e078c located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a382f7d0 @browser=#<Watir::HTMLElement:0x3298a7b0570e078c located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a382f708 @root_element=#<Watir::HTMLElement:0x46c2adda212c2d90 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a382d9a8 @browser=#<Watir::HTMLElement:0x46c2adda212c2d90 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a382d868 @root_element=#<Watir::HTMLElement:0x..fa1b463673e862ca0 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a382c008 @browser=#<Watir::HTMLElement:0x..fa1b463673e862ca0 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a3827d78 @root_element=#<Watir::HTMLElement:0x1b3a563bf3d598d4 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a38243f8 @browser=#<Watir::HTMLElement:0x1b3a563bf3d598d4 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a38243a8 @root_element=#<Watir::HTMLElement:0x..fe8607fd44c2361d4 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a381dd78 @browser=#<Watir::HTMLElement:0x..fe8607fd44c2361d4 located=true selector={element: (webdriver element)}>>>, #<StoreProperties:0x007fc3a381dd50 @root_element=#<Watir::HTMLElement:0x679f6ff12dd47760 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a381cd38 @browser=#<Watir::HTMLElement:0x679f6ff12dd47760 located=true selector={element: (webdriver element)}>>>]> to all be shoppable

         object at index 0 failed to match:
            expected `#<StoreProperties:0x007fc3a30393f0 @root_element=#<Watir::HTMLElement:0x..fb80df99fbb14eb76 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a3032fc8 @browser=#<Watir::HTMLElement:0x..fb80df99fbb14eb76 located=true selector={element: (webdriver element)}>>>.shoppable?` to return true, got false

         object at index 1 failed to match:
            expected `#<StoreProperties:0x007fc3a3032f78 @root_element=#<Watir::HTMLElement:0x12ce9696a8cd3664 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a3031cb8 @browser=#<Watir::HTMLElement:0x12ce9696a8cd3664 located=true selector={element: (webdriver element)}>>>.shoppable?` to return true, got false

         object at index 2 failed to match:
            expected `#<StoreProperties:0x007fc3a3031c90 @root_element=#<Watir::HTMLElement:0x6a325a24f6ce0974 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a3031060 @browser=#<Watir::HTMLElement:0x6a325a24f6ce0974 located=true selector={element: (webdriver element)}>>>.shoppable?` to return true, got false

         object at index 4 failed to match:
            expected `#<StoreProperties:0x007fc3a30304d0 @root_element=#<Watir::HTMLElement:0x..f9ef1392241821c78 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a301b0a8 @browser=#<Watir::HTMLElement:0x..f9ef1392241821c78 located=true selector={element: (webdriver element)}>>>.shoppable?` to return true, got false

         object at index 5 failed to match:
            expected `#<StoreProperties:0x007fc3a301b080 @root_element=#<Watir::HTMLElement:0xdf0f6eb486a7010 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a3019140 @browser=#<Watir::HTMLElement:0xdf0f6eb486a7010 located=true selector={element: (webdriver element)}>>>.shoppable?` to return true, got false

         object at index 6 failed to match:
            expected `#<StoreProperties:0x007fc3a3019118 @root_element=#<Watir::HTMLElement:0x3298a7b0570e078c located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a382f7d0 @browser=#<Watir::HTMLElement:0x3298a7b0570e078c located=true selector={element: (webdriver element)}>>>.shoppable?` to return true, got false

         object at index 7 failed to match:
            expected `#<StoreProperties:0x007fc3a382f708 @root_element=#<Watir::HTMLElement:0x46c2adda212c2d90 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a382d9a8 @browser=#<Watir::HTMLElement:0x46c2adda212c2d90 located=true selector={element: (webdriver element)}>>>.shoppable?` to return true, got false

         object at index 8 failed to match:
            expected `#<StoreProperties:0x007fc3a382d868 @root_element=#<Watir::HTMLElement:0x..fa1b463673e862ca0 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a382c008 @browser=#<Watir::HTMLElement:0x..fa1b463673e862ca0 located=true selector={element: (webdriver element)}>>>.shoppable?` to return true, got false

         object at index 11 failed to match:
            expected `#<StoreProperties:0x007fc3a381dd50 @root_element=#<Watir::HTMLElement:0x679f6ff12dd47760 located=true selector={element: (webdriver element)}>, @browser=#<Watir::Browser:0x418a2e6d85641148 url="https://shop.loblaws.ca/store-locator" title="Store Locator | Loblaws">, @platform=#<PageObject::Platforms::WatirWebDriver::PageObject:0x007fc3a381cd38 @browser=#<Watir::HTMLElement:0x679f6ff12dd47760 located=true selector={element: (webdriver element)}>>>.shoppable?` to return true, got false (RSpec::Expectations::ExpectationNotMetError)
      features/tests/tmp.feature:10:in `And all stores should be shoppable'

Failing Scenarios:
cucumber features/tests/tmp.feature:6

2 scenarios (1 failed, 1 passed)
5 steps (1 failed, 4 passed)
0m18.394s
```

**Comment by tk8817** [comment]
@daneandersen , @cheezy - if there is a bug to be fixed here it would be my pleasure to post a code-bounty or send a donation to work toward it being fixed. I am eternally grateful for the page-object gem and try my best to give back any way possible.

**Comment by AlanLGuy** [comment]
This is normal behavior of Xpath, not really a page object bug.

The "//" preceding your Xpath expression signifies "anywhere in the document" meaning it will search for a match anywhere, not necessarily within the "root" of your section. To prevent this add a "." Before the "//" this will search for a match anywhere within your page-section, rather than within the entire html document.

So:
h5(:select, xpath: '//h5[@class="store-actions shop-or-browse-store trigger-select-store"]')

Becomes:
h5( :select, xpath: './/h5[@class="store-actions shop-or-browse-store trigger-select-store"]')

**Comment by tk8817** [comment]
Thank you Alan!!! I am so glad to hear this is just my in-experience rather than a real bug.

**Comment by AlanLGuy** [comment]
Hehe. I only know because I have the good luck to work with @daneandersen :P. I ran into this issue back before page-sections existed when we were essentially doing the same things with Widgets.

It's definitely both what you'd expect the behavior to be unless you really know Xpath(since page-object does it for you for every other selector). I've used it for evil before to purposefully access elements outside the page section, but I wouldn't recommended doing it because bless it's called out in a comment you'd likely not realize that's what was going on.
